<robot name="fish_hpurv" xmlns:xacro="http://www.ros.org/wiki/xacro">

  <link name="link_base">
    <inertial>
      <origin xyz="0.0325815 0.0493006 -0.0204279" rpy="1.57 0 0"/>
      <mass value="12"/>  <!-- assume base mass remains 12 kg -->
      <inertia ixx="0.01" ixy="0" ixz="0" iyy="0.01" iyz="0" izz="0.01"/>
    </inertial>
    <visual>
      <origin xyz="0.0325815 0.0493006 0.02" rpy="1.57 0 0"/>
      <geometry>
        <mesh filename="$(find fish_hpurv)/meshes/link_base.stl"/>
      </geometry>
      <material name="link_base_material">
        <color rgba="0.796078 0.823529 0.937255 1.0"/>
      </material>
    </visual>
    <collision>
      <origin xyz="0.0325815 0.0493006 -0.0204279" rpy="1.57 0 0"/>
      <geometry>
        <mesh filename="$(find fish_hpurv)/meshes/link_base.stl"/>
      </geometry>
    </collision>
  </link>


  <link name="link_caudal">
    <inertial>
      <origin xyz="-0.378805 0 -0.05" rpy="1.5708 0 0"/>
      <mass value="0.2"/>
      <inertia ixx="0.005" ixy="0" ixz="0" iyy="0.005" iyz="0" izz="0.005"/>
    </inertial>
    <visual>
      <origin xyz="-0.378805 0 -0.05" rpy="1.5708 0 0"/>
      <geometry>
        <mesh filename="$(find fish_hpurv)/meshes/link_caudal.stl"/>
      </geometry>
      <material name="link_caudal_material">
        <color rgba="0.796078 0.823529 0.937255 1.0"/>
      </material>
    </visual>
    <collision>
      <origin xyz="-0.378805 0 -0.05" rpy="1.5708 0 0"/>
      <geometry>
        <mesh filename="$(find fish_hpurv)/meshes/link_caudal.stl"/>
      </geometry>
    </collision>
  </link>


  <joint name="caudal_joint" type="revolute">
    <origin xyz="0.425323 0.05 0.06" rpy="0 0 0"/>
    <parent link="link_base"/>
    <child link="link_caudal"/>
    <axis xyz="0 0 1"/>
    <limit effort="10" velocity="10" lower="-0.523599" upper="0.523599"/>
  </joint>


  <link name="link_pectoral_left">
    <inertial>
      <origin xyz="0.0829833 -0.0470575 0.13" rpy="3.14 0 0"/>
      <mass value="0.1"/>
      <inertia ixx="0.001" ixy="0" ixz="0" iyy="0.001" iyz="0" izz="0.001"/>
    </inertial>
    <visual>
      <origin xyz="0.0829833 -0.0470575 0.13" rpy="3.14 0 0"/>
      <geometry>
        <mesh filename="$(find fish_hpurv)/meshes/link_pectoral_left.stl"/>
      </geometry>
      <material name="link_pectoral_left_material">
        <color rgba="0.796078 0.823529 0.937255 1.0"/>
      </material>
    </visual>
    <collision>
      <origin xyz="0.0829833 -0.0470575 0.13" rpy="3.14 0 0"/>
      <geometry>
        <mesh filename="$(find fish_hpurv)/meshes/link_pectoral_left.stl"/>
      </geometry>
    </collision>
  </link>

  
  <joint name="joint1" type="revolute">
    <origin xyz="-0.05 0.18 0.035" rpy="1.57 0 0"/>
    <parent link="link_base"/>
    <child link="link_pectoral_left"/>
    <axis xyz="0 0 1"/>
    <limit effort="10" velocity="10" lower="-0.785398" upper="0.785398"/>
  </joint>


  <link name="link_pectoral_right">
    <inertial>
      <origin xyz="0.0829833 0.0470575 0.13" rpy="0 0 0"/>
      <mass value="0.1"/>
      <inertia ixx="0.001" ixy="0" ixz="0" iyy="0.001" iyz="0" izz="0.001"/>
    </inertial>
    <visual>
      <origin xyz="0.0829833 0.0470575 0.13" rpy="0 0 0"/>
      <geometry>
        <mesh filename="$(find fish_hpurv)/meshes/link_pectoral_right.stl"/>
      </geometry>
      <material name="link_pectoral_right_material">
        <color rgba="0.796078 0.823529 0.937255 1.0"/>
      </material>
    </visual>
    <collision>
      <origin xyz="0.0829833 0.0470575 0.13" rpy="0 0 0"/>
      <geometry>
        <mesh filename="$(find fish_hpurv)/meshes/link_pectoral_right.stl"/>
      </geometry>
    </collision>
  </link>

  <joint name="joint2" type="revolute">
    <origin xyz="-0.05 -0.08 0.035" rpy="-1.57 0 0"/>
    <parent link="link_base"/>
    <child link="link_pectoral_right"/>
    <axis xyz="0 0 1"/>
    <limit effort="10" velocity="10" lower="-0.785398" upper="0.785398"/>
  </joint>

  
  <gazebo reference="link_base">
    <material>Gazebo/Red</material>
    <sensor name="laser" type="ray">
      <pose>0 0 0 0 0 0</pose>
      <visualize>true</visualize>
      <update_rate>10</update_rate>
      <ray>
        <scan>
          <horizontal>
            <samples>360</samples>
            <min_angle>-3.14</min_angle>
            <max_angle>3.14</max_angle>
          </horizontal>
        </scan>
        <range>
          <min>0.3</min>
          <max>12</max>
        </range>
      </ray>
      <plugin name="laser_controller" filename="libgazebo_ros_laser.so">
        <ros>
          <argument>~/out:=/scan</argument>
        </ros>
        <output_type>sensor_msgs/LaserScan</output_type>
        <frame_name>laser_frame</frame_name>
      </plugin>
    </sensor>
  </gazebo>

  <gazebo reference="link_base">
    <sensor name="imu_sensor" type="imu">
      <update_rate>50</update_rate>
      <always_on>true</always_on>
      <visualize>true</visualize>
      <plugin name="imu_plugin" filename="libgazebo_ros_imu.so">
        <ros>
          <argument>~/out:=/imu/data</argument>
        </ros>
        <topicName>/imu/data</topicName>
        <frameName>imu_link</frameName>
      </plugin>
    </sensor>
  </gazebo>

  <gazebo reference="link_base">
    <sensor name="dvl_sensor" type="ray">
      <pose>0.03 -0.085 -0.06  0  3.14 0</pose>
      <visualize>true</visualize>
      <update_rate>10</update_rate>
      <ray>
        <scan>
          <horizontal>
            <samples>4</samples>
            <min_angle>0</min_angle>
            <max_angle>6.28319</max_angle>
          </horizontal>
          <vertical>
            <samples>1</samples>
            <min_angle>1.2217</min_angle>
            <max_angle>1.2217</max_angle>
          </vertical>
        </scan>
        <range>
          <min>0.1</min>
          <max>100</max>
        </range>
      </ray>
      <plugin name="custom_dvl_plugin" filename="libfish_dvl_plugin.so"/>
    </sensor>
  </gazebo>

  <gazebo>
  <plugin name="fish_buoyancy_plugin" filename="/home/uav/ros2_ws/install/fish_hpurv/lib/fish_hpurv/libFishBuoyancyPlugin.so">
    <fluid_density>1000</fluid_density>
    <link name="link_base">
      <volume>0.0120</volume>
      <center_of_volume>0.0325815 0.0493006 0.02</center_of_volume>
    </link>
    <link name="link_caudal">
      <volume>0.0002</volume>
      <center_of_volume>-0.378805 0 -0.05</center_of_volume>
    </link>
    <link name="link_pectoral_left">
      <volume>0.0001</volume>
      <center_of_volume>0.0829833 -0.0470575 0.13</center_of_volume>
    </link>
    <link name="link_pectoral_right">
      <volume>0.0001</volume>
      <center_of_volume>0.0829833 0.0470575 0.13</center_of_volume>
    </link>
  </plugin>
  <plugin name="caudal_force_plugin" filename="/home/uav/ros2_ws/install/fish_hpurv/lib/fish_hpurv/libCaudalForcePlugin.so"/>
</gazebo>

  
</robot>